server/lib/chat/chat-service.test.ts

```diff
@@ -0,0 +1,1505 @@
+    msgId: 'MESSAGE_ID',
+    userId: user2.id,
+    userName: user2.name,
+    channelId: testChannel.id,
+    sent: new Date('2023-03-11T00:00:00.000Z'),
+    data: {
+      type: ServerChatMessageType.JoinChannel,
+    },
+  }
+
+  // A helper method that mocks everything correctly to "join" a user to a channel.
+  const joinUserToChannel = async (
+    user: SbUser,
+    channel: ChannelInfo,
+    userChannelEntry: UserChannelEntry,
+    channelMessage: ChatMessage,
+  ) => {
+    asMockedFunction(findUserById).mockResolvedValue(user)
+    asMockedFunction(findChannelByName).mockResolvedValue(undefined)
+    asMockedFunction(createChannel).mockResolvedValue(channel)
+    asMockedFunction(addUserToChannel).mockResolvedValue(userChannelEntry)
+    asMockedFunction(addMessageToChannel).mockResolvedValue(channelMessage)
+
+    await chatService.joinChannel(testChannel.name, user1.id)
+  }
+
+  const USER1_CLIENT_ID = 'USER1_CLIENT_ID'
+  const USER2_CLIENT_ID = 'USER2_CLIENT_ID'
+
+  let client1: InspectableNydusClient
+  let client2: InspectableNydusClient
+
+  beforeEach(() => {
+    nydus = createFakeNydusServer()
+    const sessionLookup = new RequestSessionLookup()
+    const userSocketsManager = new UserSocketsManager(nydus, sessionLookup, async () => {})
+    const publisher = new TypedPublisher(nydus)
+
+    chatService = new ChatService(publisher, userSocketsManager)
+    connector = new NydusConnector(nydus, sessionLookup)
+
+    client1 = connector.connectClient(user1, USER1_CLIENT_ID)
+    client2 = connector.connectClient(user2, USER2_CLIENT_ID)
+
+    jest.clearAllMocks()
+    clearTestLogs(nydus)
+  })
+
+  describe('joinInitialChannel', () => {
+    const addUserToChannelMock = asMockedFunction(addUserToChannel)
+    const addMessageToChannelMock = asMockedFunction(addMessageToChannel)
+
+    beforeEach(async () => {
+      // NOTE(2Pac): We pre-join user2 to the channel so we can check it sees correct websocket
+      // messages from user1 joining initially.
+      await joinUserToChannel(
+        user2,
+        shieldBatteryChannel,
+        user2ShieldBatteryChannelEntry,
+        joinUser2ShieldBatteryChannelMessage,
+      )
+
+      asMockedFunction(findUserById).mockResolvedValue(user1)
+      asMockedFunction(getChannelInfo).mockResolvedValue([shieldBatteryChannel])
+      addUserToChannelMock.mockResolvedValue(user1ShieldBatteryChannelEntry)
+      addMessageToChannelMock.mockResolvedValue(joinUser1ShieldBatteryChannelMessage)
+    })
+
+    test("should throw if user doesn't exist", async () => {
+      asMockedFunction(findUserById).mockResolvedValue(undefined)
+
+      await expect(
+        chatService.joinInitialChannel(user1.id, dbClient, Promise.resolve()),
+      ).rejects.toThrowErrorMatchingInlineSnapshot(`"User doesn't exist"`)
+    })
+
+    test('works when user exists', async () => {
+      await chatService.joinInitialChannel(user1.id, dbClient, Promise.resolve())
```

Comment:
It would be nice to ensure ordering for the things that can only happen after the transaction has been committed (that is, you should pass a `Deferred` for the promise here, and check that those things only happen after you resolve it)